### PR TITLE
25 error component

### PR DESCRIPTION
**Pull request by summer-cook**
# summary
- configures errors to display differently in dev vs prod environments
- adds the error component throughout the app
- separates the configuration & helper functions into two separate files

# related
[initial error component#25](https://github.com/scientist-softserv/webstore-component-library/issues/25)

# video

home page error
  - only will err on featured services component and not full page
  - allows you to dismiss the error
& multiple errors
  - example is the request page
  - take the full page and have a button to return to your previous page
  
https://share.getcloudapp.com/z8uldRP6

# screenshot 
single error (in prod)
image

single error (in dev)
image

# notes
- we would most likely need a global state in order to use a single error component at the top level of the app. 
- looked into react ErrorBoundary but its possible that there is a 3rd party library that could achieve this

### CODE REVIEW

**Review by labradford — APPROVED** [review]

File: utils/api/helpers.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
+
```
Comment (alishaevn) [inline utils/api/helpers.js:1]: can the content of this file be moved to "utils/helpers.js"?

**Review by alishaevn — CHANGES_REQUESTED** [review]

File: pages/browse/index.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import { useEffect, useState } from 'react'
 import { useRouter } from 'next/router'
-import { Item, SearchBar } from '@scientist-softserv/webstore-component-library'
-import { configureServices, useFilteredWares } from '../../utils'
+import { Item, Error, SearchBar } from '@scientist-softserv/webstore-component-library'
+import { configureServices, configureErrors, useFilteredWares } from '../../utils'
```
Comment (alishaevn) [inline pages/browse/index.js:4]: alphabetize the imports please.

File: pages/browse/index.js [hunk 3]
```diff
@@ -19,7 +19,11 @@ const Browse = () => {
     return router.push({ pathname: '/browse', query: { q: value } }, (value.length > 0 ? `/browse?q=${value}` : '/browse'))
   }
 
-  if (isError) return <h1>Error...</h1>
+  if (isError) {
+    return (
+      <Error errors={configureErrors([isError])} router={router} />
+    )
+  }
```
Comment (alishaevn) [inline pages/browse/index.js:26]:
```suggestion
  if (isError) return <Error errors={configureErrors([isError])} router={router} />
```

File: pages/requests/index.js [hunk 4]
```diff
@@ -1,18 +1,25 @@
...
+import { useRouter } from 'next/router'
 import {
+  Error,
   LinkedButton,
   Loading,
   RequestList,
   Title,
 } from '@scientist-softserv/webstore-component-library'
-import { dark, useAllRequests } from '../../utils'
+import { dark, useAllRequests, configureErrors } from '../../utils'
```
Comment (alishaevn) [inline pages/requests/index.js:10]: alphabetize the imports please.

File: utils/api/configurations.js [hunk 5]
```diff
@@ -53,84 +60,43 @@ export const configureRequests = ({ data, path }) => {
...
-            interval = seconds
-            intervalType = 'second'
-          }
-        }
+// takes an array of errors for each page or component
+export const configureErrors = (errors) => {
+  const env = process.env.NODE_ENV
+  let errorText = []
+  let errorTitle = ''
+  if (env == 'development') {
```
Comment (alishaevn) [inline utils/api/configurations.js:68]:
```suggestion
 
   if (env === 'development') {
```

File: utils/api/configurations.js [hunk 6]
```diff
@@ -53,84 +60,43 @@ export const configureRequests = ({ data, path }) => {
...
+    errors.map(error => {
+      errorText.push(JSON.stringify(error))
```
Comment (alishaevn) [inline utils/api/configurations.js:70]: does this behave as you expect it to when the errors array includes an empty hash?

File: utils/api/configurations.js [hunk 7]
```diff
@@ -53,84 +60,43 @@ export const configureRequests = ({ data, path }) => {
...
+      errorTitle = errors.length > 1 ? 'There were multiple errors.' : error.name
```
Comment (alishaevn) [inline utils/api/configurations.js:71]: if `errors = [{ title: 'some', message: 'data' }, {}]`, then we only have 1 error and not multiple. this code doesn't account for that scenario as is.

File: utils/api/configurations.js [hunk 8]
```diff
@@ -53,84 +60,43 @@ export const configureRequests = ({ data, path }) => {
...
+      return {
+        errorText, errorTitle
       }
+    })
```
Comment (alishaevn) [inline utils/api/configurations.js:75]:
```suggestion
      return { errorText, errorTitle }
    })

```

File: utils/api/configurations.js [hunk 9]
```diff
@@ -53,84 +60,43 @@ export const configureRequests = ({ data, path }) => {
...
+    return {
+      errorText: errorText, // returns an array that will be mapped over in the component
+      errorTitle: errorTitle,
+      variant: 'warning'
     }
   }
-
-  if (interval > 1 || interval === 0) {
-    intervalType += 's'
+  if (env == 'production'){
```
Comment (alishaevn) [inline utils/api/configurations.js:82]:
```suggestion

    if (env === 'production') {
```

File: utils/api/configurations.js [hunk 10]
```diff
@@ -53,84 +60,43 @@ export const configureRequests = ({ data, path }) => {
...
+    // could add different variants for each status code
+    let text
+    switch (errors[0].response.status) {
+    case 404:
+      text = 'This page or section was not found.'
+      break
+    
+    default:
+      text = "We're working on getting this back up and running as soon as possible."
+    }
```
Comment (alishaevn) [inline utils/api/configurations.js:93]:
```suggestion
    }

```

File: pages/requests/index.js [hunk 11]
```diff
@@ -1,18 +1,25 @@
...
   const { requests, isLoading, isError } = useAllRequests()
   const { user, userError, userLoading } = props
 
-  if (isError) return <h1>{`${isError.name}: ${isError.message}`}</h1>
-  if (userError) return <h1>{`${userError.name}: ${userError.message}`}</h1>
+  if (isError || userError ) {
+    return (
+      <Error errors={configureErrors([isError, userError])} router={router} />
+    )
+  }
```
Comment (alishaevn) [inline pages/requests/index.js:21]: can you refactor to a single line?

File: pages/services/[ware].js [hunk 12]
```diff
@@ -1,13 +1,17 @@
...
   const router = useRouter()
   const { id } = router.query
   const { ware, isLoading, isError } = useOneWare(id)
 
-  if (isError) return <h1>{`${isError.name}: ${isError.message}`}</h1>
+  if (isError) {
+    return (
+      <Error errors={configureErrors([isError])} router={router} />
+    )
+  }
```
Comment (alishaevn) [inline pages/services/[ware].js:14]: can you refactor to a single line?

**Review by alishaevn — CHANGES_REQUESTED** [review]

**Review by alishaevn — APPROVED** [review]

**Pull request merged**